### DISCUSSION TITLE
Using Tailwind CSS 3 in Marp

### DISCUSSION THREAD

**Discussion by sbibauw** [marp-team/marp#237] [Q&A] [answered]
In #192, @yhatt proposed to use Tailwind CSS in Marp to implement certain styling options, such as columns.
The possibilities of Tailwind are impressive within Marp and would really make it a very powerful utility for anyone wanting more customization of styles with HTML within Markdown.

However, with the new Tailwind CSS, I no longer am able to import Tailwind with an import declaration in the style frontmatter (`style: @import url('https://cdn.tailwindcss.com/3.0.0');` doesn't work). Any thoughts on how to do this?

Or, @yhatt, if other people find it useful, any chance of actually bundling Tailwind within Marp? 

Besides, if some people have good examples of uses of Tailwind within Marp, I would be very interested in seeing them!

**Comment by yhatt** [comment, marked as answer]
> For Tailwind CSS v3.0, the CSS-based CDN build we’ve offered in the past has been replaced by the new Play CDN, which gives you the full power of the new engine right in the browser with no build step.
>
> _— https://tailwindcss.com/docs/upgrade-guide#play-cdn_

If you want to use Tailwind v3, just replacing CSS `@import` to `<script>` tag should work.

```diff
 ---
 marp: true
-style: @import url('https://unpkg.com/tailwindcss@^2/dist/utilities.min.css');
 ---

+<script src="https://cdn.tailwindcss.com/3.0.0"></script>
```

:warning: Please keep in mind Marp team does not so recommend to enable HTML in Markdown for saving users from any malicious things. Especially putting `<script>` tag(s) can run powerful scripts in the rendered slides, but _we do never take responsibility whatever happens._ VS Code's Markdown preview has disabled running any scripts by default.

**Reply by sbibauw** [reply]
It works.
And it works in VSCode Preview after changing the `Markdown: Change Preview Security Settings` to "Disable (not recommended)".

For some reason, activating Tailwind v3 seems to remove bullet points and indentation in the exported HTML, but I'm seeing you're updating the styles in #235 so I suppose that might get fixed :-)

**Reply by yhatt** [reply]
A good to know thing is that Tailwind CSS has the opinionated base styles called [Preflight](https://tailwindcss.com/docs/preflight). This will remove basic styles for list elements: https://tailwindcss.com/docs/preflight#lists-are-unstyled

However, *the most of Marp themes have expected default styles provided by browser and not compatible with Tailwind preflight.* The previous CSS import had imported only utility classes (`utlities.css`) instead of full CSS, to avoid reset of styles by preflight.

To disable preflight in v3 Play CDN, add extra `<script>` tag to set Tailwind config [for disabling preflight](https://tailwindcss.com/docs/preflight#disabling-preflight).

```markdown
---
marp: true
---

<script src="https://cdn.tailwindcss.com/3.0.0"></script>
<script>tailwind.config = { corePlugins: { preflight: false } }</script>
```

> #235 is an update for our website https://marp.app/ and has no effect to Marp users.

**Reply by mil-ad** [reply]
This works great with HTML but when I export to PDF I can see the snippet on the first page and tailwind doesn't load.

**Comment by yhatt** [comment]
> Or, @yhatt, if other people find it useful, any chance of actually bundling Tailwind within Marp?

I strongly believe _keeping a project minimum is important for the ecosystem sustainability_, so I'm following [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) while developing Marp. All stylings Tailwind CSS can do are available also in the plain CSS so I'm not going to work for the full integration with Tailwind CSS for now.

---

As mentioned also in #192, the developer who think it is useful can try to add integration with Tailwind through Marp plugin: By overloading [protected `renderStyle` method of an instance extended from `Marpit`](https://github.com/marp-team/marpit/blob/2d2e9fc64b2b07cec1a5bb45d2d81dbbc5cef722/src/marpit.js#L220-L231), the plugin can add a mutation logic before and after Marp style rendering. In fact, Marp Core is exactly doing the same thing to Marpit framework to add postprocess for rendered style. (https://github.com/marp-team/marp-core/blob/c1f8ccdc8185a80003a04528b0c8f22880fe50eb/src/marp.ts#L102-L115)

A below snippet is a minimum Marp plugin example to add the mutation logic for adding Tailwind transformation via PostCSS. (Not confirmed whether works)

```javascript
import plugin from '@marp-team/marpit/plugin'
import postcss from 'postcss'
import tailwindcss from 'tailwindcss'

export const marpTailwindPlugin = plugin(({ marpit }) => {
  // Overload Marpit#renderStyle
  const { renderStyle } = marpit

  marpit.renderStyle = (...args) => {
    const originalOutput = renderStyle(...args)
    return postcss([tailwindcss]).process(originalOutput).css
  }
})
```

If it had been published to npm as like as `marp-plugin-tailwind` package, Marp CLI users could recieve and evaluate the power of Tailwind via [the customized engine](https://github.com/marp-team/marp-cli#functional-engine) easily.

```javascript
// engine.js for Marp CLI
const { marpTailwindPlugin } = require('marp-plugin-tailwind')
module.exports = ({ marp }) => marp.use(marpTailwindPlugin)
```

**Reply by sbibauw** [reply]
Thanks a lot. I understand your decision to keep it minimal, and if it is as simple as adding a `<script>` line, then it's not really needed.
I'm not familiar enough with Javascript to implement a plugin, but it's great to know that the possibility exists.

**Reply by mil-ad** [reply]
Thanks for sharing the snippet. It gives me this error:

```
[  INFO ] Converting 1 markdown...
  marp-cli [Worker 1] Start processing ... +0ms
  marp-cli [Worker 1] Processing /home/milad/Documents/presentations/foo/foo.md ... +0ms
  marp-cli Converting foo.md ... +0ms
  marp-cli [Worker 2] Start processing ... +0ms
  marp-cli [Worker 2] Finish processing. +0ms
  marp-cli [Worker 3] Start processing ... +0ms
  marp-cli [Worker 3] Finish processing. +0ms
  marp-cli [Worker 4] Start processing ... +0ms
  marp-cli [Worker 4] Finish processing. +0ms
  marp-cli [Worker 5] Start processing ... +0ms
  marp-cli [Worker 5] Finish processing. +0ms
You are using marp-custom-engine.
[  WARN ] Cannot register theme CSS: themes/compiled/base/catppuccin-latte.css (Marpit theme CSS requires @theme meta.)
  marp-cli CLIError: Failed converting Markdown. (Cannot read properties of undefined (reading 'themeSet'))
  marp-cli     at he (/home/milad/Documents/presentations/foo/node_modules/@marp-team/marp-cli/lib/marp-cli-zRIl56xr.js:1:8268)
  marp-cli     at Kc (/home/milad/Documents/presentations/foo/node_modules/@marp-team/marp-cli/lib/marp-cli-zRIl56xr.js:48:195553) {
  marp-cli   errorCode: 1
  marp-cli } +160ms
  ```
